### DISCUSSION TITLE
Example to setup and control robot with ros

### DISCUSSION THREAD

**Discussion by rua0ra1** [cyberbotics/wrestling#31] [Q&A] [answered]
Hi,
I am more familiar and comfortable  with ROS1. I have seen couple of example to setup and control robot with ROS2. I am wondering whether are there any example to setup and control with ROS1. And also I am very new to Docker images and confused how to change docker file to interface with ROS1. I really appreciate if anyone can answer this or share related resource to help me to setup ROS1.
Thank you

**Comment by omichel** [comment, marked as answer]
Basically, it would work in the same way as the ROS 2 example, but instead of choosing a `humble-simulation` base image for your Dockerfile, you should choose a noetic base image, like [noetic-ros-base](https://hub.docker.com/layers/library/ros/noetic-ros-base/images/sha256-a03e2f0efd75d1be2f26643ca32f10143761ffcfb655a2a5dfcfffddde1c82c5?context=explore) to which you will have add the [webots_ros](http://wiki.ros.org/webots_ros) package with `apt-get install ros-noetic-webots-ros` in your Dockerfile.

See https://hub.docker.com/_/ros for more details about ROS 1 & 2 docker images and how to use them.

See also [this tutorial](https://www.cyberbotics.com/doc/guide/tutorial-9-using-ros) on how to use ROS 1 with Webots.

**Reply by rua0ra1** [reply]
Thank you @omichel . I will go through above  listed  links.

**Reply by rua0ra1** [reply]
Hi @omichel ,
I updated my OS to ubuntu 22. As I am planning to use ROS2 to control the robot and it also helps me to improve my ros2 skills. I installed ros2, webots 2023a and webots_ros2 packages. Now I am trying to run "python3 -u participant.py". But I am getting the following issues 
"Cannot connect to Webots instance, retrying in 2 second..."

I really appreciate it if you can help me to solve this issue.

**Reply by omichel** [reply]
I order to connect your controller as an extern controller, you should also edit the world file and change the [controller of your robot](https://github.com/cyberbotics/wrestling/blob/main/worlds/wrestling.wbt#L118) from `"participant"` to `"<extern>"`.

**Reply by rua0ra1** [reply]
Sorry @omichel , it didn't resolve the issue. I can launch all the webots_ros2 examples (for example "ros2 launch webots_ros2_epuck robot_launch.py") without any issues. But,I am getting the same  error for both "python3 -u participant.py" and also "ros2 launch paricipant robot_launch.py" as 
"Cannot connect to Webots instance, retrying in 2 second...".
I have added webots_home env variable to my bash file. Do I need to another variables ?.

following are the launch.log file output

676889893.0875871 [INFO] [launch]: All log files can be found below /home/ram/.ros/log/2023-02-20-21-44-53-048074-Astra-14191
1676889893.0879903 [INFO] [launch]: Default logging verbosity is set to INFO
1676889893.2015324 [INFO] [driver-1]: process started with pid [14193]
1676889944.2251849 [ERROR] [driver-1]: process has died [pid 14193, exit code 1, cmd '/opt/ros/humble/lib/webots_ros2_driver/driver --ros-args --params-file /tmp/launch_params_oa2nfujt']. Hope it would help you troubleshoot the issue

**Reply by Jean-Eudes-le-retour** [reply]
Here is a small guide to launch the ROS example controller locally:

1. Clone the [ROS example controller](https://github.com/cyberbotics/wrestling-alice-ros-2)
2. Open the `world/wrestling.wbt` file with Webots
3. In a new terminal, navigate to the folder of the `wrestling-alice-ros-2` repository
4. Run the following commands: 
``` bash
cd controllers/participant/
colcon build
. /opt/ros/humble/setup.sh
. ./install/setup.sh
ros2 launch participant robot_launch.py
```

The controller should connect to Webots and make the robot move its arm. You might need to click on the play button.

Can you tell us if this works for you?

**Reply by rua0ra1** [reply]
Finally It is working :) :)...... Thank you so much @Jean-Eudes-le-retour  and @omichel  for all your help and support
